File: vripper-server/src/main/java/tn/mnlr/vripper/tasks/LeaveThanksRunnable.java
package tn.mnlr.vripper.tasks;

import lombok.extern.slf4j.Slf4j;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.protocol.HttpClientContext;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import tn.mnlr.vripper.SpringContext;
import tn.mnlr.vripper.Utils;
import tn.mnlr.vripper.jpa.domain.LogEvent;
import tn.mnlr.vripper.jpa.domain.Post;
import tn.mnlr.vripper.jpa.repositories.ILogEventRepository;
import tn.mnlr.vripper.services.ConnectionService;
import tn.mnlr.vripper.services.DataService;
import tn.mnlr.vripper.services.SettingsService;

import java.io.UnsupportedEncodingException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Slf4j
public class LeaveThanksRunnable implements Runnable {

  private final ConnectionService cm;
  private final DataService dataService;
  private final HttpClientContext context;
  private final SettingsService settingsService;
  private final ILogEventRepository eventRepository;
  private final boolean authenticated;
  private final Post post;
  private final LogEvent logEvent;

  public LeaveThanksRunnable(Post post, boolean authenticated, HttpClientContext context) {
    this.post = post;
    this.authenticated = authenticated;
    this.context = context;
    cm = SpringContext.getBean(ConnectionService.class);
    dataService = SpringContext.getBean(DataService.class);
    eventRepository = SpringContext.getBean(ILogEventRepository.class);
    settingsService = SpringContext.getBean(SettingsService.class);
    logEvent =
        new LogEvent(
            LogEvent.Type.THANKS,
            LogEvent.Status.PENDING,
            LocalDateTime.now(),
            String.format("Leaving thanks for %s", post.getUrl()));
    eventRepository.save(logEvent);
  }

  @Override
  public void run() {
    try {
      logEvent.setStatus(LogEvent.Status.PROCESSING);
      eventRepository.update(logEvent);
      if (!settingsService.getSettings().getVLogin()) {
        logEvent.setMessage(
            String.format(
                "Will not send a like for %s\nAuthentication with ViperGirls option is disabled",
                post.getUrl()));
        logEvent.setStatus(LogEvent.Status.DONE);
        eventRepository.update(logEvent);
        return;
      }
      if (!settingsService.getSettings().getVThanks()) {
        logEvent.setMessage(
            String.format(
                "Will not send a like for %s\nLeave thanks option is disabled", post.getUrl()));
        logEvent.setStatus(LogEvent.Status.DONE);
        eventRepository.update(logEvent);
        return;
      }
      if (!authenticated) {
        logEvent.setMessage(
            String.format("Will not send a like for %s\nYou are not authenticated", post.getUrl()));
        logEvent.setStatus(LogEvent.Status.ERROR);
        eventRepository.update(logEvent);
        return;
      }
      if (post.isThanked()) {
        logEvent.setMessage(
            String.format("Will not send a like for %s\nAlready left a thanks", post.getUrl()));
        logEvent.setStatus(LogEvent.Status.DONE);
        eventRepository.update(logEvent);
        return;
      }

      HttpPost postThanks =
          cm.buildHttpPost(settingsService.getSettings().getVProxy() + "/post_thanks.php", null);
      List<NameValuePair> params = new ArrayList<>();
      params.add(new BasicNameValuePair("do", "post_thanks_add"));
      params.add(new BasicNameValuePair("using_ajax", "1"));
      params.add(new BasicNameValuePair("p", post.getPostId()));
      params.add(new BasicNameValuePair("securitytoken", post.getSecurityToken()));
      try {
        postThanks.setEntity(new UrlEncodedFormEntity(params));
      } catch (UnsupportedEncodingException e) {
        String error = String.format("Request error for %s", post.getUrl());
        log.error(error, e);
        logEvent.setMessage(error + "\n" + Utils.throwableToString(e));
        logEvent.setStatus(LogEvent.Status.ERROR);
        eventRepository.update(logEvent);
        return;
      }

      postThanks.addHeader("Referer", settingsService.getSettings().getVProxy());
      postThanks.addHeader(
          "Host",
          settingsService.getSettings().getVProxy().replace("https://", "").replace("http://", ""));

      CloseableHttpClient client = cm.getClient().build();

      try (CloseableHttpResponse response = client.execute(postThanks, context)) {
        if (response.getStatusLine().getStatusCode() / 100 == 2) {
          post.setThanked(true);
          dataService.updatePostThanks(post.isThanked(), post.getId());
        }
        EntityUtils.consumeQuietly(response.getEntity());
      }
      logEvent.setStatus(LogEvent.Status.DONE);
      eventRepository.update(logEvent);
    } catch (Exception e) {
      String error = String.format("Failed to leave a thanks for %s", post.getUrl());
      log.error(error, e);
      logEvent.setMessage(error + "\n" + Utils.throwableToString(e));
      logEvent.setStatus(LogEvent.Status.ERROR);
      eventRepository.update(logEvent);
    }
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/tasks/LeaveThanksRunnable.kt
package me.mnlr.vripper.tasks

import me.mnlr.vripper.delegate.LoggerDelegate
import me.mnlr.vripper.entities.LogEvent
import me.mnlr.vripper.entities.LogEvent.Status.*
import me.mnlr.vripper.entities.Post
import me.mnlr.vripper.formatToString
import me.mnlr.vripper.repositories.LogEventRepository
import me.mnlr.vripper.services.HTTPService
import me.mnlr.vripper.services.SettingsService
import org.apache.http.NameValuePair
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpPost
import org.apache.http.client.protocol.HttpClientContext
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.message.BasicNameValuePair
import org.apache.http.util.EntityUtils
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import java.io.UnsupportedEncodingException

class LeaveThanksRunnable(
    private val post: Post,
    private val authenticated: Boolean,
    private val context: HttpClientContext
) : KoinComponent, Runnable {
    private val log by LoggerDelegate()
    private val cm: HTTPService by inject()
    private val settingsService: SettingsService by inject()
    private val eventRepository: LogEventRepository by inject()
    private val logEvent: LogEvent

    init {
        logEvent = eventRepository.save(
            LogEvent(
                type = LogEvent.Type.THANKS,
                status = PENDING,
                message = "Leaving thanks for $post.url"
            )
        )
    }

    override fun run() {
        try {
            eventRepository.update(logEvent.copy(status = PROCESSING))
            if (!settingsService.settings.viperSettings.login) {
                eventRepository.update(
                    logEvent.copy(
                        status = DONE,
                        message = "Will not send a like for ${post.url}\nAuthentication with ViperGirls option is disabled"
                    )
                )
                return
            }
            if (!settingsService.settings.viperSettings.thanks) {
                eventRepository.update(
                    logEvent.copy(
                        status = DONE,
                        message = "Will not send a like for ${post.url}\nLeave thanks option is disabled"
                    )
                )
                return
            }
            if (!authenticated) {
                eventRepository.update(
                    logEvent.copy(
                        status = ERROR,
                        message = "Will not send a like for ${post.url}\nYou are not authenticated"
                    )
                )
                return
            }
            val postThanks: HttpPost = cm.buildHttpPost(
                "${settingsService.settings.viperSettings.host}/post_thanks.php", HttpClientContext.create()
            )
            val params: MutableList<NameValuePair> = ArrayList()
            params.add(BasicNameValuePair("do", "post_thanks_add"))
            params.add(BasicNameValuePair("using_ajax", "1"))
            params.add(BasicNameValuePair("p", post.postId))
            params.add(BasicNameValuePair("securitytoken", post.token))
            try {
                postThanks.entity = UrlEncodedFormEntity(params)
            } catch (e: UnsupportedEncodingException) {
                val error = "Request error for ${post.url}"
                log.error(error, e)
                eventRepository.update(
                    logEvent.copy(
                        status = ERROR, message = """
                    $error
                    ${e.formatToString()}
                    """.trimIndent()
                    )
                )
                return
            }
            postThanks.addHeader("Referer", settingsService.settings.viperSettings.host)
            postThanks.addHeader(
                "Host", settingsService.settings.viperSettings.host.replace("https://", "").replace("http://", "")
            )
            val client: CloseableHttpClient = cm.client.build()
            client.execute(postThanks, context).use { response ->
                try {
                } finally {
                    EntityUtils.consumeQuietly(response.entity)
                }
            }
            eventRepository.update(logEvent.copy(status = DONE))
        } catch (e: Exception) {
            val error = "Failed to leave a thanks for $post"
            log.error(error, e)
            eventRepository.update(
                logEvent.copy(
                    status = ERROR, message = """
                $error
                ${e.formatToString()}
                """.trimIndent()
                )
            )
        }
    }
}